vulcanai/models/utils.py

```diff
@@ -4,13 +4,77 @@
 import torch.nn as nn
 import torch.nn.functional as F
 from torch.autograd import Variable
+from torch.utils.data import DataLoader, TensorDataset
 
 import numpy as np
 import math
 from sklearn.metrics import confusion_matrix
 from sklearn.preprocessing import LabelBinarizer
 from collections import OrderedDict as odict
+from collections import defaultdict
 
+def get_probs(network, loader, index_to_iter, ls_feat_vals):
+    """
+    Returns probability for each object within loader based on output from training neural network
+    Parameters:
+        network : vulcan.model
+	    training vulcan network
+	loader : torch.dataloader
+	    dataloader containing validation set
+	index_to_iter : string
+	    feature to iterate through adjusting values
+	ls_feat_vals : list
+	    values to iterate through for feature in index_to_iter
+    Returns:
+	dct_scores : dictionary
+	    dictionary of scores
+    """
+    dct_scores = defaultdict()
+    for index in range(len(loader)):
+        dct_scores[index] = {}
+    for index in range(len(loader)):
```

Comment:
Hey @joemehltretter as I've been thinking in terms of optimization, this whole function could probably done in terms of whole-matrix operations, instead of by individual subject. Right now it's sending a single subject at a time through the forward pass, when really we can (and should) send all of them at once. Essentially we would just change the value for all the subjects, and collect all the results, just as we normally would. Also, all the processing like rounding should probably be done at the end, outside of this function, again for efficiencies sake and so we don't have any surprises. If you want me to tackle that then please let me know, I'd be happy to try although I'd love your feedback as I'm not as familiar with bootfold.

Reply:
As long as we ensure we can identify the best value for each "subject" then I'm fine with doing matrix operations. This was the easiest way we found in v1 to ensure that we look at an individual subject. Feel free to do it I don't mind at all.